compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,432 @@
+
+            return;
+        }
+        self.super_place(place, ctxt, loc);
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum Value<'tcx> {
+    /// Not a pointer, or we can't know.
+    Unknown,
+    /// We know the value to be a pointer to this place.
+    Pointer(Place<'tcx>),
+}
+
+/// For each local, save the place corresponding to `*local`.
+#[instrument(level = "trace", skip(tcx, body))]
+fn compute_local_values<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &Body<'tcx>,
+    ssa: &SsaLocals,
+    fully_replacable_locals: &BitSet<Local>,
+) -> IndexVec<Local, Value<'tcx>> {
+    // Returns true iff we can use `p` as a pointee.
+    let is_constant = |p: Place<'_>| {
+        // We only allow `Deref` as the first projection, to avoid surprises.
+        if p.projection.first() == Some(&PlaceElem::Deref) {
+            // `p == (*some_local).xxx`, it is constant only if `some_local` is constant.
+            // We approximate constness using SSAness.
+            matches!(ssa.assignments[p.local], Set1::One(_))
+                && p.projection[1..].iter().all(|e| match e {
+                    PlaceElem::Field(..)
+                    | PlaceElem::ConstantIndex { .. }
+                    | PlaceElem::Subslice { .. }
+                    | PlaceElem::Downcast(..)
+                    | PlaceElem::OpaqueCast(_) => true,
+                    PlaceElem::Deref | PlaceElem::Index(_) => false,
+                })
+        } else {
+            p.projection.iter().all(|e| match e {
+                PlaceElem::Field(..)
+                | PlaceElem::ConstantIndex { .. }
+                | PlaceElem::Subslice { .. }
+                | PlaceElem::Downcast(..)
+                | PlaceElem::OpaqueCast(_) => true,
+                PlaceElem::Deref | PlaceElem::Index(_) => false,
+            })
+        }
+    };
+
+    let mut targets = IndexVec::from_elem(Value::Unknown, &body.local_decls);
+
+    for &local in &ssa.assignment_order {
+        debug!(?local);
+
+        // Only visit if we have something to do.
+        let Value::Unknown = targets[local] else { bug!() };
+
+        // This is not SSA: mark that we don't know the value.
+        debug!(assignments = ?ssa.assignments[local]);
+        let Set1::One(LocationExtended::Plain(loc)) = ssa.assignments[local] else {
+            targets[local] = Value::Unknown;
+            continue;
+        };
+
+        // If this a mutable reference that we cannot fully replace, mark it as unknown.
+        if body.local_decls[local].ty.is_mutable_ptr() && !fully_replacable_locals.contains(local) {
+            targets[local] = Value::Unknown;
+            continue;
+        }
+
+        // `loc` must point to a direct assignment to `local`.
+        let Either::Left(stmt) = body.stmt_at(loc) else { bug!() };
+        let Some((_target, rvalue)) = stmt.kind.as_assign() else { bug!() };
+        assert_eq!(_target.as_local(), Some(local));
+
+        let rvalue = match rvalue {
```

Comment:
do the `targets[local]` assignments inside and skip all the `Value::Unknown` cases